File: skills/udf-gen-test/templates/java/src/test/java/com/udf/UnitTest.java
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf;

import java.util.Arrays;
import java.util.List;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

public class UnitTest {

    private static SparkSession spark;
    private static ClassLoader origContextClassLoader;

    @BeforeClass
    public static void setUp() {
        origContextClassLoader = TestUtils.installMutableClassLoader();
        spark = SparkSession.builder()
            .appName("UDF Unit Test")
            .master("local[*]")
            .config("spark.plugins", "com.nvidia.spark.SQLPlugin")
            .config("spark.rapids.skipGpuArchitectureCheck", "true")
            .config("spark.rapids.sql.mode", "explainOnly")
            .config("spark.sql.adaptive.enabled", "false")
            .enableHiveSupport()
            .getOrCreate();
    }

    @AfterClass
    public static void tearDown() {
        if (spark != null) spark.stop();
        if (origContextClassLoader != null) {
            Thread.currentThread().setContextClassLoader(origContextClassLoader);
        }
    }

    /**
     * TODO: Create a test DataFrame with diverse test cases including edge cases.
     *
     * Example:
     * <pre>{@code
     *   StructType schema = new StructType(new StructField[]{
     *       DataTypes.createStructField("id", DataTypes.IntegerType, false),
     *       DataTypes.createStructField("credit_score", DataTypes.IntegerType, true)
     *   });
     *   List<Row> data = Arrays.asList(
     *       RowFactory.create(1, 800),
     *       RowFactory.create(2, 550),
     *       RowFactory.create(3, null)
     *   );
     *   return spark.createDataFrame(data, schema);
     * }</pre>
     */
    public static Dataset<Row> createTestData(SparkSession spark) {
        return null; // TODO
    }

    /**
     * TODO: Register the UDF with Spark.
     *
     * Example (Hive UDF):
     * <pre>{@code
     *   spark.sql("CREATE TEMPORARY FUNCTION " + udfName
     *       + " AS 'com.udf.CalculateRiskUDF'");
     * }</pre>
     */
    public static void registerUDF(SparkSession spark, String udfName) {
        // TODO
    }

    /**
     * TODO: Execute the UDF on the test DataFrame and return the result.
     *
     * Example:
     * <pre>{@code
     *   testDF.createOrReplaceTempView("test_table");
     *   return spark.sql("SELECT *, " + udfName
     *       + "(credit_score) AS risk_level FROM test_table");
     * }</pre>
     */
    public static Dataset<Row> executeUDF(SparkSession spark, String udfName, Dataset<Row> testDF) {
        return null; // TODO
    }

    /**
     * TODO: Verify UDF results using Assert statements.
     *
     * Example:
     * <pre>{@code
     *   Row[] results = (Row[]) resultDF.sort("id").collect();
     *   Assert.assertEquals("LOW", results[0].getAs("risk_level"));
     *   Assert.assertEquals("MEDIUM", results[1].getAs("risk_level"));
     *   Assert.assertEquals("UNKNOWN", results[2].getAs("risk_level"));
     * }</pre>
     */
    public static void verifyUDFResults(Dataset<Row> resultDF, Dataset<Row> testDF) {
        // TODO
    }

    @Test
    public void testUDFProducesCorrectResults() {
        Dataset<Row> testDF = createTestData(spark).repartition(1);

        registerUDF(spark, "placeholder_udf_name");
        Dataset<Row> resultDF = executeUDF(spark, "placeholder_udf_name", testDF);

        verifyUDFResults(resultDF, testDF);
    }
}

File: skills/udf-gen-test/templates/scala/src/test/scala/com/udf/UnitTest.scala
/*
 * SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package com.udf

import org.apache.spark.sql.{DataFrame, Row, SparkSession}
import org.apache.spark.sql.types._
import org.scalatest.Assertions
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.BeforeAndAfterAll

object UnitTest extends Assertions {
  /**
   * TODO: Create a test DataFrame with diverse test cases including edge cases
   * (at least 10+ cases).
   *
   * Example:
   * {{{
   *   val schema = StructType(Seq(
   *     StructField("id", IntegerType, nullable = false),
   *     StructField("credit_score", IntegerType, nullable = true)
   *   ))
   *   val testData = Seq(
   *     Row(1, 800),
   *     Row(2, 550),
   *     Row(3, null)
   *     // ...
   *   )
   *   spark.createDataFrame(spark.sparkContext.parallelize(testData), schema)
   * }}}
   */
  def createTestData(spark: SparkSession): DataFrame = ???

  /**
   * TODO: Register the UDF with Spark.
   *
   * Examples:
   * {{{
   *   spark.udf.register(udfName, new CalculateRiskUDF())   // Scala UDF
   *   spark.udf.register(udfName, new FormatPhoneUDF(), StringType)   // Java UDF
   *   spark.sql(s"CREATE TEMPORARY FUNCTION $udfName AS 'com.udf.IntegerMultiplyBy2UDF'")   // Hive UDF
   * }}}
   */
  def registerUDF(spark: SparkSession, udfName: String): Unit = ???

  /**
   * TODO: Execute the UDF on the test DataFrame and return the result.
   *
   * Example:
   * {{{
   *   testDF.createOrReplaceTempView("test_table")
   *   spark.sql(s"SELECT *, $udfName(credit_score) AS risk_level FROM test_table")
   * }}}
   */
  def executeUDF(spark: SparkSession, udfName: String, testDF: DataFrame): DataFrame = ???

  /**
   * TODO: Assert the UDF results match expectations.
   *
   * Example:
   * {{{
   *   val results = resultDF.collect().sortBy(_.getAs[Int]("id"))
   *   assert(results(0).getAs[String]("risk_level") === "LOW")
   *   assert(results(1).getAs[String]("risk_level") === "MEDIUM")
   *   assert(results(2).getAs[String]("risk_level") === "UNKNOWN")
   * }}}
   */
  def assertUDFResults(resultDF: DataFrame, testDF: DataFrame): Unit = ???
}

class UnitTest extends AnyFunSuite with BeforeAndAfterAll {

  var spark: SparkSession = _

  override def beforeAll(): Unit = {
    spark = SparkSession.builder()
      .appName("UDF Unit Test")
      .master("local[4]")
      .config("spark.plugins", "com.nvidia.spark.SQLPlugin")
      .config("spark.rapids.skipGpuArchitectureCheck", "true")
      .config("spark.rapids.sql.mode", "explainOnly")
      .config("spark.sql.adaptive.enabled", "false")
      .enableHiveSupport()
      .getOrCreate()
  }

  override def afterAll(): Unit = {
    if (spark != null) spark.stop()
  }

  test("UDF produces correct results") {
    // Repartition down to 2 tasks to ensure we exercise multi-row columns.
    val testDF = UnitTest.createTestData(spark).repartition(2)

    UnitTest.registerUDF(spark, "placeholder_udf_name")
    val resultDF = UnitTest.executeUDF(spark, "placeholder_udf_name", testDF)

    UnitTest.assertUDFResults(resultDF, testDF)
  }
}
